server/src/service.rs

```diff
@@ -99,9 +99,41 @@ impl Service {
     #[zbus(out_args("locked", "prompt"))]
     pub async fn lock(
         &mut self,
-        _objects: Vec<OwnedObjectPath>,
+        objects: Vec<OwnedObjectPath>,
+        #[zbus(signal_emitter)] signal_emitter: SignalEmitter<'_>,
     ) -> Result<(Vec<OwnedObjectPath>, ObjectPath), ServiceError> {
-        todo!()
+        let mut locked: Vec<OwnedObjectPath> = Vec::new();
+        let prompt = ObjectPath::default();
+        let collections = self.collections.lock().await;
+        let manager = self.manager.lock().await;
+
+        for object in &objects {
+            for collection in collections.iter() {
+                if object == collection.path() || collection.items().await.contains(object) {
+                    for item in collection.inner_items().await {
+                        if !item.is_locked().await {
+                            item.set_locked(true).await;
+                            let signal_emitter = manager.signal_emitter(item.path().clone())?;
+                            item.locked_changed(&signal_emitter).await?;
+                        }
+                        let signal_emitter = manager.signal_emitter(collection.path().clone())?;
+                        Collection::item_changed(&signal_emitter, item.path()).await?;
```

Comment:
This is not really what I meant. Instead, make a new method in impl Item, called `changed` that would get the signal emitter from the service manager, and call `Collection::item_changed` and pass the correct params to it. Same for 
`Service::collection_changed`.
You can then call that method inside set_locked so that locking/unlocking only calls that without having to deal with all the details

Reply:
Good idea, Thanks! I have named the new method as `locked_updated` because we have other properties. If you still think `changed` is more appropriate here, I'm happy to change this.

Reply:
Oh! found another issue in my impl. Fixing this now.

Reply:
No, the changed is for the item_changed, it has nothing to do with locked. Anything specific to locked state, should be called in set_locked, no extra methods.

Also the impl is not exactly what I meant, you want to have Service::collection_changed, in a `async fn changed(&self)` for Collection type as well as the same thing for Item. No passing state of a Collection to Item.

Reply:
> Oh! found another issue in my impl. Fixing this now.

Fixed the [issue](https://github.com/bilelmoussaoui/oo7/pull/150#discussion_r1826894735)